Automattic/woocommerce-payments
# Show UTC label in date time column header of transactions CSV

Fixes https://github.com/Automattic/woocommerce-payments/issues/8522

Fixes https://github.com/Automattic/woocommerce-payments-server/issues/2847

I had a hard time digesting the problem here so summarizing it here:
- There are 3 timezones at play here: browser's timezone (in my case UTC+7), store's timezone (set at General > Settings > Timezone), and server's (UTC).
- The behavior on base branch is as follow:
  - I'll use an example of a transaction I created at `2024-09-19 06:11` in my local time (UTC+7).
  - I set my store's timezone to be UTC+1.
  - On server, transaction time is stored as UTC, `2024-09-18 23:11`.
  - When plugin fetches the transaction from server, it will receive the time in UTC.
  - [This Javascript code shows the time in store's locale](https://github.com/Automattic/woocommerce-payments/blob/c83f5de040a7b846bb5f15ac87009b98bf9cae44/client/transactions/list/index.tsx#L467-L470), ie in store's timezone. Read [`dateI18n`'s description](https://github.com/WordPress/gutenberg/blob/trunk/packages/date/src/index.js#L523). Therefore, I see the transaction time on the Transactions list page as `2024-09-19 00:11`.
  - I can download CSV in 2 different ways based on the length of data.
    - If it's just one page, I will download the CSV right away. This is called 'browser' download. The times displayed in the CSV are in UTC.
    - Otherwise, the server will email me the CSV file. This is called 'endpoint' download. The times displayed in the CSV are in UTC.

Problem:
- When I 'browser' download, in the CSV, the header says `Date / Time` and it's not clear what the timezone is.
- The header on transaction list page is `Date / Time` and it's not clear what the timezone is. *This is not within the scope of this PR*. Created a separate issue: https://github.com/Automattic/woocommerce-payments/issues/9472.

#### Changes proposed in this Pull Request

- Since times in CSV are in UTC, add `(UTC)` to the date time column header. Apply the same for blocked transactions. This PR will add UTC for 'browser' download only because CSV from 'endpoint' download already has UTC in the Date / Time header.

![Screenshot 2024-09-19 at 03 56 36](https://github.com/user-attachments/assets/07e5ae2d-eae5-44ac-bde9-4e8826cc58f8)

I do not add UTC to the other list pages (deposits and disputes) because they are showing dates only (without time).

#### Testing instructions

* Set your blog's timezone to non-UTC. Go to wp-admin > Settings > General > Timezone.
* `npm run build:client`.
* Go to wp-admin > Payments > Transactions. Click download.

* Go to wp-admin > Payments > Transactions > Blocked. Click download. I don't have any blocked transactions, so the download button does not appear for me.

* Test the 2 types of download: from browser and from server.
  - Modify this [line](https://github.com/Automattic/woocommerce-payments/blob/542e27f93037c23007799c7b9b020b12e7c84a93/client/transactions/list/index.tsx#L720).

To force it using the `browser` method:
```
const downloadType = 1 > rows.length ? 'endpoint' : 'browser';
```

To force it using the `endpoint` method:
```
const downloadType = 1000000 > rows.length ? 'endpoint' : 'browser';
```

Confirm that the date time header column is `Date / Time (UTC)`. For CSV coming from server, if no email arrives at your inbox, check the mail logger installed in your local server WordPress.

-------------------


**Post merge**
